I got that working by adding `setattr(cls, bases[0]._fieldsname, d[bases[0]._fieldsname])` (a bit of a hack, I know)

But the following still fails:
```python
from ndx_speech import Transcription
from pynwb.epoch import TimeIntervals
import pandas as pd

words = TimeIntervals.from_dataframe(pd.DataFrame(
    {'start_time': [.1, 2.], 'stop_time': [.8, 2.3], 'label': ['hello', 'there']}), name='words')

tt = Transcription(words=words)
```
with the following error:

```python
Traceback (most recent call last):
  File "/Users/bendichter/dev/ndx/ndx-speech/src/pynwb/ndx_speech/test/test.py", line 8, in <module>
    tt = Transcription(words=words)
  File "/Users/bendichter/dev/hdmf/src/hdmf/utils.py", line 388, in func_call
    return func(self, **parsed['args'])
  File "/Users/bendichter/dev/hdmf/src/hdmf/build/map.py", line 1426, in __init__
    arg_val.parent = self
  File "/Users/bendichter/dev/hdmf/src/hdmf/container.py", line 86, in parent
    raise Exception('cannot reassign parent')
Exception: cannot reassign parent
```

--- a/src/hdmf/build/map.py
+++ b/src/hdmf/build/map.py
@@ -1421,9 +1421,6 @@ class TypeMap(object):
                 arg_val = kwargs.get(f, None)
                 if arg_val is not None:
                     setattr(self, f, arg_val)
-                    if issubclass(type(arg_val), Container):
-                        arg_val.parent = self
-                        # there is some interaction here with fields['child'] = True
 
         return {'__init__': __init__, base._fieldsname: tuple(fields)}
 

--- a/src/hdmf/build/map.py
+++ b/src/hdmf/build/map.py
@@ -1335,15 +1335,14 @@ class TypeMap(object):
         'isodatetime': datetime
     }
 
-    def get_container_type(self, container_name):
+    def __get_container_type(self, container_name):
         container_type = None
         for val in self.__container_types.values():
             container_type = val.get(container_name)
             if container_type is not None:
                 return container_type
         if container_type is None:
-            # it is not clear when this would ever happen
-            return container_name
+            raise TypeDoesNotExistError("Type '%s' does not exist." % container_name)
 
     def __get_type(self, spec):
         if isinstance(spec, AttributeSpec):
@@ -1361,9 +1360,9 @@ class TypeMap(object):
         if isinstance(spec, LinkSpec):
             return Container
         if spec.data_type_def is not None:
-            return self.get_container_type(spec.data_type_def)
+            return self.__get_container_type(spec.data_type_def)
         if spec.data_type_inc is not None:
-            return self.get_container_type(spec.data_type_inc)
+            return self.__get_container_type(spec.data_type_inc)
         if spec.shape is None and spec.dims is None:
             return self._type_map.get(spec.dtype)
         return 'array_data', 'data'
@@ -1465,7 +1464,13 @@ class TypeMap(object):
             for k, field_spec in attr_names.items():
                 if not spec.is_inherited_spec(field_spec):
                     fields[k] = field_spec
-            d = self.__get_cls_dict(parent_cls, fields, spec.name)
+            try:
+                d = self.__get_cls_dict(parent_cls, fields, spec.name)
+            except TypeDoesNotExistError as e:
+                name = spec.get('data_type_def', 'Unknown')
+                raise ValueError("Cannot dynamically generate class for type '%s'. " % name
+                                 + str(e)
+                                 + " Please define that type before defining '%s'." % name)
             cls = ExtenderMeta(str(name), bases, d)
             self.register_container_type(namespace, data_type, cls)
         return cls
@@ -1673,3 +1678,6 @@ class TypeMap(object):
             raise ValueError('No ObjectMapper found for container of type %s' % str(container.__class__.__name__))
         else:
             return attr_map.get_builder_name(container)
+
+class TypeDoesNotExistError(Exception):
+    pass

--- a/tests/unit/test_io_hdf5_h5tools.py
+++ b/tests/unit/test_io_hdf5_h5tools.py
@@ -3,6 +3,7 @@ import unittest2 as unittest
 import tempfile
 import warnings
 import numpy as np
+import re
 
 from hdmf.utils import docval, getargs
 from hdmf.data_utils import DataChunkIterator
@@ -560,6 +561,35 @@ class TestCacheSpec(unittest.TestCase):
                 if type(foo_obj) is Baz1:
                     read_baz = foo_obj
             self.assertEqual(read_baz.baz2.attr1, baz2.attr1)
+
+    def test_cache_spec_get_class_in_class_wrong_order(self):
+        spec_catalog = self.manager.namespace_catalog.get_namespace(CORE_NAMESPACE).catalog
+        foo_spec = spec_catalog.get_spec('Foo')
+        # Baz1 class contains an object of Baz2 class
+        baz_spec2 = GroupSpec('A composition inside',
+                              data_type_def='Baz2',
+                              data_type_inc=foo_spec,
+                              attributes=[
+                                  AttributeSpec('attr3', 'an example float attribute', 'float'),
+                                  AttributeSpec('attr4', 'another example float attribute', 'float')])
+
+        baz_spec1 = GroupSpec('A composition test outside',
+                              data_type_def='Baz1',
+                              data_type_inc=foo_spec,
+                              attributes=[AttributeSpec('attr3', 'an example float attribute', 'float'),
+                                          AttributeSpec('attr4', 'another example float attribute', 'float')],
+                              groups=[GroupSpec('A composition inside', data_type_inc='Baz2')])
+
+        # add directly into the existing spec_catalog. would not do this normally.
+        spec_catalog.register_spec(baz_spec1, 'test.yaml')
+        spec_catalog.register_spec(baz_spec2, 'test.yaml')
+
+
+        # Setup all the data we need
+        msg = ("Cannot dynamically generate class for type 'Baz1'. Type 'Baz2' does not exist. "
+               "Please define that type before defining 'Baz1'.")
+        with self.assertRaisesRegex(ValueError, re.escape(msg)):
+            Baz1 = self.manager.type_map.get_container_cls(CORE_NAMESPACE, 'Baz1')
 
     def __get_types(self, catalog):
         types = set()

